src/circuit/gadget/sinsemilla/commit_ivk.rs

```diff
@@ -0,0 +1,802 @@
+
+            let b = b_0.zip(b_1).zip(b_2).map(|((b_0, b_1), b_2)| {
+                let b1_shifted = b_1 * pallas::Base::from_u64(1 << 4);
+                let b2_shifted = b_2 * pallas::Base::from_u64(1 << 5);
+                b_0 + b1_shifted + b2_shifted
+            });
+
+            // Constrain b_0 to be 4 bits.
+            let b_0 = self.sinsemilla_config.lookup_config.witness_short_check(
+                layouter.namespace(|| "b_0 is 4 bits"),
+                b_0,
+                4,
+            )?;
+            // Constrain b_2 to be 5 bits.
+            let b_2 = self.sinsemilla_config.lookup_config.witness_short_check(
+                layouter.namespace(|| "b_2 is 5 bits"),
+                b_2,
+                5,
+            )?;
+            // b_1 will be boolean-constrained in the custom gate.
+
+            let b = MessagePiece::from_field_elem(
+                sinsemilla_chip.clone(),
+                layouter.namespace(|| "b = b_0 || b_1 || b_2"),
+                b,
+                1,
+            )?;
+
+            (b_0, b_1, b_2, b)
+        };
+
+        // c = bits 5..=244 of `nk`
+        let c = {
+            let c = nk.value().map(|value| bitrange_subset(value, 5..245));
+            MessagePiece::from_field_elem(
+                sinsemilla_chip.clone(),
+                layouter.namespace(|| "c"),
+                c,
+                24,
+            )?
+        };
+
+        // `d = d_0||d_1` = (bits 245..=253 of `nk`) || (bit 254 of `nk`)
+        let (d_0, d_1, d) = {
+            let d_0 = nk.value().map(|value| bitrange_subset(value, 245..254));
+            let d_1 = nk.value().map(|value| bitrange_subset(value, 254..255));
+
+            let d = d_0
+                .zip(d_1)
+                .map(|(d_0, d_1)| d_0 + d_1 * pallas::Base::from_u64(1 << 9));
+
+            // Constrain d_0 to be 9 bits.
+            let d_0 = self.sinsemilla_config.lookup_config.witness_short_check(
+                layouter.namespace(|| "d_0 is 9 bits"),
+                d_0,
+                9,
+            )?;
+            // d_1 will be boolean-constrained in the custom gate.
+
+            let d = MessagePiece::from_field_elem(
+                sinsemilla_chip.clone(),
+                layouter.namespace(|| "c = d_0 || d_1"),
+                d,
+                1,
+            )?;
+
+            (d_0, d_1, d)
+        };
+
+        let (ivk, zs) = {
+            let message = Message::from_pieces(
+                sinsemilla_chip.clone(),
+                vec![a.clone(), b.clone(), c.clone(), d.clone()],
+            );
+            let domain = CommitDomain::new(
+                sinsemilla_chip,
+                ecc_chip,
+                &SinsemillaCommitDomains::CommitIvk,
+            );
+            domain.short_commit(layouter.namespace(|| "Hash ak||nk"), message, rivk)?
+        };
+
+        let z13_a = zs[0][13];
+        let z14_c = zs[2][14];
```

Comment:
If I'm interpreting the circuit correctly (in particular the comment in `nk_canonicity`), this is a constraint bug:
```suggestion
        let z13_c = zs[2][13];
```
That is, if d<sub>1</sub> = 1, we want to constrain c to 130 bits, and then constrain b<sub>2</sub> + c*2<sup>5</sup> (which will be 135 bits) to be < t<sub>P</sub>. The tests (which use valid inputs) pass because if z<sub>c,13</sub> = 0 then z<sub>c,14</sub> will always be zero.

I suspect the bug occurred because for the b<sub>2</sub> + c*2<sup>5</sup> constraint, we _do_ need to select z<sub>14</sub>.